presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java

```diff
@@ -0,0 +1,552 @@
+                            pages = ImmutableList.of();
+                        }
+                    }
+
+                    if (shouldAcknowledge && acknowledgePages) {
+                        // Acknowledge token without handling the response.
+                        // The next request will also make sure the token is acknowledged.
+                        // This is to fast release the pages on the buffer side.
+                        resultClient.acknowledgeResultsAsync(result.getNextToken());
+                    }
+
+                    for (SerializedPage page : pages) {
+                        if (!isChecksumValid(page)) {
+                            throw new PrestoException(SERIALIZED_PAGE_CHECKSUM_ERROR, format("Received corrupted serialized page from host %s", HostAddress.fromUri(uri)));
+                        }
+                    }
+
+                    // add pages:
+                    // addPages must be called regardless of whether pages is an empty list because
+                    // clientCallback can keep stats of requests and responses. For example, it may
+                    // keep track of how often a client returns empty response and adjust request
+                    // frequency or buffer size.
+                    if (clientCallback.addPages(PageBufferClient.this, pages)) {
+                        pagesReceived.addAndGet(pages.size());
+                        rowsReceived.addAndGet(pages.stream().mapToLong(SerializedPage::getPositionCount).sum());
+                    }
+                    else {
+                        pagesRejected.addAndGet(pages.size());
+                        rowsRejected.addAndGet(pages.stream().mapToLong(SerializedPage::getPositionCount).sum());
+                    }
+                }
+                catch (PrestoException e) {
+                    handleFailure(e, resultFuture);
+                    return;
+                }
+
+                synchronized (PageBufferClient.this) {
+                    // client is complete, acknowledge it by sending it a delete in the next request
+                    if (result.isClientComplete()) {
+                        completed = true;
+                    }
+                    if (future == resultFuture) {
+                        future = null;
+                    }
+                    lastUpdate = DateTime.now();
+                }
+                requestsCompleted.incrementAndGet();
+                clientCallback.requestComplete(PageBufferClient.this);
+            }
+
+            @Override
+            public void onFailure(Throwable t)
+            {
+                log.debug("Request to %s failed %s", uri, t);
+                checkNotHoldsLock(this);
+
+                t = resultClient.rewriteException(t);
+                if (!(t instanceof PrestoException) && backoff.failure()) {
+                    String message = format("%s (%s - %s failures, failure duration %s, total failed request time %s)",
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format long: &#39;getFailureCount&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 363. Consider formatting this data with &#39;%d&#39; instead.